### PR TITLE
Support "move if to guard" with an else branch

### LINKED ISSUE #10997: "move if to guard" should work even if there's an else
image

This should suggest moving if conditon to the guard (duplicating the pattern). Relevant code:

https://github.com/rust-analyzer/rust-analyzer/blob/db2a7087b994e20f264f26ad6db75184282ad120/crates/ide_assists/src/handlers/move_guard.rs#L121-L123

### PR DESCRIPTION
**Pull request by weirane**
Support the assist `move_arm_cond_to_match_guard` when there is an else branch.

I have two questions:
1. How to indent the first line of a match arm? `matcharm.indent()` doesn't seem to work. so I hard coded four spaces here:
https://github.com/weirane/rust-analyzer/blob/95a0de85d5d659e876e731725475f1c022719f31/crates/ide_assists/src/handlers/move_guard.rs#L162-L163

2. I find a little issue in the original implementation, this code
```rust
let y = match 92 {
    x => {
        if x == 0 {$0
            false
        }
    }
    _ => true,
};
```
will be transformed to
```rust
let y = match 92 {
    x if x == 0 => false
    _ => true,
};
```
a comma is missing after the `false`. Should I also fix that? Or this can go in a separate PR.

Closes #10997.

### CODE REVIEW

File: crates/ide_assists/src/handlers/move_guard.rs [hunk 1]
```diff
@@ -149,6 +146,32 @@ pub(crate) fn move_arm_cond_to_match_guard(acc: &mut Assists, ctx: &AssistContex
...
+                edit.insert(then_arm_end, format!("\n{}{} => ", spaces, match_pat));
+                match &else_block.tail_expr() {
+                    Some(else_expr) if else_only_expr => {
+                        edit.insert(then_arm_end, else_expr.syntax().text());
+                        edit.insert(then_arm_end, ",");
+                    }
+                    _ if replace_node != *if_expr.syntax() => {
+                        edit.insert(then_arm_end, else_block.dedent(1.into()).syntax().text());
+                    }
+                    _ => edit.insert(then_arm_end, else_block.syntax().text()),
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/move_guard.rs:172]: can you add some `cov_mark::check!`s for the three branches here?

Basically adding a `cov_mark::hit!(move_guard_ifelse_expr_only);` to the first arm, `cov_mark::check!(move_guard_ifelse_expr_only);` to one of the tests that hits this match arm, same for the other two with a different identifier in the macro.

That way we can assert that those tests test these arms properly.

**Comment by Veykril** [comment]
> Should I also fix that? Or this can go in a separate PR.

That can be fixed in a follow up PR.

> How to indent the first line of a match arm?

That seems odd if it doesn't work, hardcoding the space is fine for now I guess. Our syntax formatting is a big mess anyways that we will hopefully replace some day.


Note we can even improve on this improvement by having this work for if else chains of any length by making each `else if` a separate block as long as the chain ends in a normal `else` block without an `if`. This can be done in a separate PR as well though.

**Comment by weirane** [comment]
I've added the `cov_mark::check!`s and I think I won't do `else if`s in this PR.

**Comment by Veykril** [comment]
Thanks!
bors r+

**Comment by lnicola** [comment]
changelog fix (first contribution) support `Move condition to guard` with an `else` branch

**Pull request merged**